### PR TITLE
Alerting: Update provisioning to validate user-defined UID on create

### LINKED ISSUE #62904: Provisioning API alert_rule validation of limits
**What happened**:
When creating provisioned alert rules using the API, if the uid for the alert exceeds the 40 char limit creation succeeds and the resulting uid is truncated. 

These truncated uid alert rules don't retain the provisioned status in the UI, It's not clear if that's a feature. 

The deletion of alert rules does not also truncate the incoming uid, which leads to an automated deletion problem. 

**What you expected to happen**:
Creation of an alert rule where the uid or other fields exceed the limits should fail

**How to reproduce it (as minimally and precisely as possible)**:
Create an alert-rule using the provisioner api with a uid exceeding 40 chars

**Anything else we need to know?**:

- Grafana version: 9.3.6

### PR DESCRIPTION
**Pull request by yuri-tceretian**
**What is this feature?**
Fixes provisioning service to validate user-assigned UID of alert rule and contact points, and return error if the UID does not meet the requirements. 

Validation does not affect the update operation and therefore, existing resources will not be affected.

**Why do we need this feature?**
To enforce uniform rules for UIDs across all APIs and databases. In SQLite mode, the long UID will be accepted because the field has TEXT type. However, in MySQL and Postgres - the UID will be rejected by the database.

Also, it will help prevent users from getting in trouble while using manually assigned UIDs. For example, UID like `my=abc` will be accepted by provisioning but user will not be able to reach out the rule via UI.

**Who is this feature for?**

[Add information on what kind of user the feature is for.]

**Which issue(s) does this PR fix?**:

Fixes https://github.com/grafana/grafana/issues/62904

**Special notes for your reviewer:**

Please check that:
- [x] It works as expected from a user's perspective.
- [x] If this is a pre-GA feature, it is behind a feature toggle.
- [x] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: conf/provisioning/alerting/sample.yaml [hunk 1]
```diff
@@ -13,7 +13,7 @@ apiVersion: 1
 #     interval: 60s
 #     # <list, required> list of rules that are part of the rule group
 #     rules:
-#       # <string, required> unique identifier for the rule
+#       # <string, required> unique identifier for the rule. Should not exceed 40 symbols. Allowed only letters, numbers, - (hyphen), and _ (underscore).
 #     - uid: my_id_1
```
Comment (brendamuir) [inline conf/provisioning/alerting/sample.yaml:17]:
```suggestion
#       # <string, required> unique identifier for the rule. Should not exceed 40 symbols. Only letters, numbers, - (hyphen), and _ (underscore) allowed.
#     - uid: my_id_1
```

File: conf/provisioning/alerting/sample.yaml [hunk 2]
```diff
@@ -82,7 +82,7 @@ apiVersion: 1
 #     # <string, required> name of the contact point
 #     name: cp_1
 #     receivers:
-#       # <string, required> unique identifier for the receiver
+#       # <string, required> unique identifier for the receiver. Should not exceed 40 symbols. Allowed only letters, numbers, - (hyphen), and _ (underscore).
 #     - uid: first_uid
```
Comment (brendamuir) [inline conf/provisioning/alerting/sample.yaml:86]:
```suggestion
#       # <string, required> unique identifier for the receiver. Should not exceed 40 symbols. Only letters, numbers, - (hyphen), and _ (underscore) allowed.
#     - uid: first_uid
```

File: docs/sources/alerting/set-up/provision-alerting-resources/file-provisioning/index.md [hunk 3]
```diff
@@ -65,7 +65,7 @@ groups:
     interval: 60s
     # <list, required> list of rules that are part of the rule group
     rules:
-      # <string, required> unique identifier for the rule
+      # <string, required> unique identifier for the rule. Should not exceed 40 symbols. Allowed only letters, numbers, - (hyphen), and _ (underscore).
```
Comment (brendamuir) [inline docs/sources/alerting/set-up/provision-alerting-resources/file-provisioning/index.md:68]:
```suggestion
      # <string, required> unique identifier for the rule. Should not exceed 40 symbols. Only letters, numbers, - (hyphen), and _ (underscore) allowed.
```

File: docs/sources/alerting/set-up/provision-alerting-resources/file-provisioning/index.md [hunk 4]
```diff
@@ -158,7 +158,7 @@ contactPoints:
     # <string, required> name of the contact point
     name: cp_1
     receivers:
-      # <string, required> unique identifier for the receiver
+      # <string, required> unique identifier for the receiver. Should not exceed 40 symbols. Allowed only letters, numbers, - (hyphen), and _ (underscore).
       - uid: first_uid
```
Comment (brendamuir) [inline docs/sources/alerting/set-up/provision-alerting-resources/file-provisioning/index.md:162]:
```suggestion
      # <string, required> unique identifier for the receiver. Should not exceed 40 symbols. Only letters, numbers, - (hyphen), and _ (underscore) allowed.
      - uid: first_uid
```

File: pkg/util/shortid_generator_test.go [hunk 5]
```diff
@@ -84,3 +85,47 @@ func TestIsShortUIDTooLong(t *testing.T) {
...
+		uid      string
+		expected error
+	}{
+		{
+			name:     "no error when string is of correct length",
+			uid:      "f8cc010c-ee72-4681-89d2-d46e1bd47d33",
+			expected: nil,
+		},
+		{
+			name:     "error when it is empty",
```
Comment (rwwiv) [inline pkg/util/shortid_generator_test.go:101]: I would change these to all read "error when _uuid_ ..."

**Review by rwwiv — APPROVED** [review]
Couple very minor nits around testing, but LGTM 🚀

File: pkg/util/shortid_generator_test.go [hunk 6]
```diff
@@ -84,3 +85,47 @@ func TestIsShortUIDTooLong(t *testing.T) {
...
+			expected: ErrUIDTooLong,
+		},
+		{
+			name:     "error when it has invalid characters",
+			uid:      "f8cc010c.ee72.4681;89d2+d46e1bd47d33",
+			expected: ErrUIDFormatInvalid,
+		},
+		{
+			name:     "error when it has whitespaces",
+			uid:      " ",
```
[the remark this answers is not in the document]
Comment (yuri-tceretian) [reply pkg/util/shortid_generator_test.go:117]: This tests asserts that a string that contains whitespaces does not pass validation

[the remark this answers is not in the document]
Comment (rwwiv) [reply pkg/util/shortid_generator_test.go:117]: New test name makes it much clearer, works for me 👍

**Review by alexweav — APPROVED** [review]
LGTM, nice approach with making it so only newly created objects uptake this, so it's not breaking.

What's here is good, though I suspect we could go further - check the PUT for contact points:
image

What happens if you PUT /contact-points/<some-UID>
with:
```
{
    ...
    "uid": "some-long-invalid-uid"
}
```
I honestly don't remember the behavior here - does it try to update the actual UID of the contact point? If so, we could potentially also attempt to validate in this case too (as long as we don't break updates if the path UID is already invalid to begin with).

**Comment by yuri-tceretian** [comment]
> What happens if you PUT /contact-points/ 

I just tested it and uid in the payload is ignored. Perhaps we need to fail the request if uid is specified.

**Pull request merged**